### PR TITLE
ADD: Import private keys in hex or base64 formats

### LINKED ISSUE #3477: Support hex & base64 for private keys for single-address wallets
in addition to WIFs.

test cases:
image

image

Private Key Hexadecimal Format (64 characters [0-9A-F]):
D556170609D43CAAC751EB81D5400425E029A86C2718CD33EC4D7D43CE1BB306
Private Key Base64 (44 characters):
1VYXBgnUPKrHUeuB1UAEJeApqGwnGM0z7E19Q84bswY=

### PR DESCRIPTION
**Pull request by ojokne**
This pr adds importing private keys in hex and base64 formats for single address wallets

closes #3477

It works by converting the hex/base64 private key to compressed  WIF ( Wallet import format)

Test cases:

Private Key in hex 
D556170609D43CAAC751EB81D5400425E029A86C2718CD33EC4D7D43CE1BB306

Private Key in  Base64
1VYXBgnUPKrHUeuB1UAEJeApqGwnGM0z7E19Q84bswY=

### CODE REVIEW

File: class/wallet-import.ts [hunk 1]
```diff
@@ -319,6 +319,24 @@ const startImport = (
     }
 
     yield { progress: 'wif' };
+
+    const isHexKey = /^[0-9a-fA-F]{64}$/.test(text);
+    const isBase64Key = /^[A-Za-z0-9+/=]{43,44}$/.test(text);
```
Comment (GladosBlueWallet) [inline class/wallet-import.ts]: `isBase64Key` is very permissive: it accepts any 43-44 chars from `[A-Za-z0-9+/=]` (including `=` in the middle) and doesn't enforce proper padding / length%4. That means lots of random strings will be treated as base64 and then potentially re-encoded to WIF if they decode to 32 bytes. Consider using a stricter base64 check (padding only at end, length multiple of 4), or decode first then re-encode and compare for round-trip sanity.

File: class/wallet-import.ts [hunk 2]
```diff
@@ -319,6 +319,24 @@ const startImport = (
...
+    if (isHexKey) {
+      rawKeyBuffer = Buffer.from(text, 'hex');
+    } else if (isBase64Key) {
+      try {
+        rawKeyBuffer = Buffer.from(text, 'base64');
+      } catch (_) {}
+    }
+
+    if (rawKeyBuffer && rawKeyBuffer.length === 32) {
+      text = wif.encode(0x80, rawKeyBuffer, true);
```
Comment (GladosBlueWallet) [inline class/wallet-import.ts]: This forces `wif.encode(0x80, rawKeyBuffer, true)` (mainnet + *compressed*=true). Two concerns: (1) a raw 32-byte key does not imply compressed vs uncompressed, and forcing compressed changes legacy P2PKH addresses; users importing a raw key might expect the uncompressed legacy address variant. (2) hardcoding 0x80 bakes in mainnet; if import flow ever supports testnet/regtest, this will silently create the wrong WIF. Maybe try both compressed flags (or gate by user/network settings), and use the network prefix from settings instead of a literal.

**Review by GladosBlueWallet — COMMENTED** [review]
I ran the key-import enrichment through my test chamber. It mostly survived. A few edge cases did not.

File: class/wallet-import.ts (hunk 2 above)
Comment (ojokne) [reply class/wallet-import.ts]: I specifically chose `0x80` to support mainnet only and made it compressed as it is recommended.

you can advice on this @Overtorment

File: class/wallet-import.ts [hunk 3]
```diff
@@ -319,6 +319,24 @@ const startImport = (
...
+    let rawKeyBuffer;
+
+    if (isHexKey) {
+      rawKeyBuffer = Buffer.from(text, 'hex');
+    } else if (isBase64Key) {
+      try {
+        rawKeyBuffer = Buffer.from(text, 'base64');
+      } catch (_) {}
+    }
+
```
Comment (limpbrains) [inline class/wallet-import.ts]: I guess `rawKeyBuffer = Buffer.from(text, 'hex');` is not under try / catch because `isHexKey` is less permissive ?

**Review by limpbrains — APPROVED** [review]

File: class/wallet-import.ts [hunk 4]
```diff
@@ -319,6 +319,24 @@ const startImport = (
     }
 
     yield { progress: 'wif' };
+
```
Comment (limpbrains) [inline class/wallet-import.ts:322]: Please add a comment here about what are you doing and why.

File: class/wallet-import.ts (hunk 3 above)
Comment (ojokne) [reply class/wallet-import.ts]: i think the try/ catch block was redundant, removed it

File: class/wallet-import.ts (hunk 4 above)
Comment (ojokne) [reply class/wallet-import.ts:322]: Added comments describing each step

**Review by limpbrains — APPROVED** [review]

File: class/wallet-import.ts [hunk 5]
```diff
@@ -319,6 +319,25 @@ const startImport = (
...
+    // check if text is in hex or base64 format
+    const isHexKey = /^[0-9a-fA-F]{64}$/.test(text);
+    const isBase64Key = /^[A-Za-z0-9+/=]{43,44}$/.test(text);
+
+    let rawKeyBuffer;
+
+    if (isHexKey) {
+      rawKeyBuffer = Buffer.from(text, 'hex');
```
Comment (Overtorment) [inline class/wallet-import.ts]: we are getting rid of Buffer, use `Uint8Array`

File: class/wallet-import.ts (hunk 2 above)
Comment (Overtorment) [reply class/wallet-import.ts]: mainnet is fine.

as for compressed flag, we should `yield` 2 different wallets, compressed and uncompressed, or, so be more specific:

* 2 different wallets, compressed and uncompressed (p2pkh)
* compressed (segwit wrapped)
* compressed (segwit native)

(segwit spec specifically mentions uncompressed pubkeys are no good for segwit)

Comment (Overtorment) [reply class/wallet-import.ts]: so, dont pass wif down the chain, `yield` several wallets instead

**Review by Overtorment — CHANGES_REQUESTED** [review]

File: class/wallet-import.ts (hunk 2 above)
Comment (ojokne) [reply class/wallet-import.ts]: I can modify and check if a used an uncompressed  p2pkh exists, then yield it, otherwise yield the compressed p2pkh, p2wpkh,  p2tr and p2wpkh-p2sh

The other option is create a new class for uncompressed p2pkh (Uncompressed LegacyWallet)  and be able to yield both compressed and uncompressed, along with p2wpkh,  p2tr and p2wpkh-p2sh  and user makes the choice.

I prefer the option of yielding uncompressed p2pkh only if there is an existing wallet that has been used

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @marcosrdz] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/marcosrdz

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @marcosrdz] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/marcosrdz

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @marcosrdz] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/marcosrdz

File: class/wallet-import.ts (hunk 5 above)
Comment (ojokne) [reply class/wallet-import.ts]: the `wif` dependency still uses buffer thats why


Screenshot From 2026-05-05 18-28-06

**Review by ncoelho — APPROVED** [review]

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @marcosrdz] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/marcosrdz

File: tests/integration/import.test.ts [hunk 6]
```diff
@@ -710,4 +711,54 @@ describe('import procedure', () => {
...
+    assert.strictEqual(store.state.wallets[0].getSecret(), 'L4NQfHXE9xQ11neMH9MMjRX5fwiT2nsQYz96LxbMJpk3UU8yUyEC');
+    assert.strictEqual(store.state.wallets[1].getSecret(), 'L4NQfHXE9xQ11neMH9MMjRX5fwiT2nsQYz96LxbMJpk3UU8yUyEC');
+    assert.strictEqual(store.state.wallets[2].getSecret(), 'L4NQfHXE9xQ11neMH9MMjRX5fwiT2nsQYz96LxbMJpk3UU8yUyEC');
+    assert.strictEqual(store.state.wallets[3].getSecret(), 'L4NQfHXE9xQ11neMH9MMjRX5fwiT2nsQYz96LxbMJpk3UU8yUyEC');
+
+    assert.strictEqual(store.state.wallets[0].getAddress(), 'bc1qxscvu3w04nj9k2eukx30897xldfygzqqkmtdav');
+    assert.strictEqual(store.state.wallets[1].getAddress(), '38z3svUJKMFj4G3rGrRzcARp2N4vdqV8NK');
+    assert.strictEqual(store.state.wallets[2].getAddress(), '15kxc4PEBR8yXqpgEkb83YijfjSkd2xrKV');
+    assert.strictEqual(store.state.wallets[3].getAddress(), 'bc1pxnmhjaug5jqm3xqz5fekme8f8dax869z96ufqyllgz5g3wpw2gpqvqxqqy');
+  });
```
Comment (Overtorment) [inline tests/integration/import.test.ts:744]: there should be 5, no?

* taproot
* segwit native
* segwit wrapped
* legacy (compressed)
* legacy (uncompressed)

Comment (Overtorment) [reply tests/integration/import.test.ts:744]: or we dont support uncompressed pubkeys in legacy wallets..? my memory is a bit fuzzy here

**Review by Overtorment — CHANGES_REQUESTED** [review]

File: tests/integration/import.test.ts (hunk 6 above)
Comment (ojokne) [reply tests/integration/import.test.ts:744]: > there should be 5, no?
> 
> * taproot
> * segwit native
> * segwit wrapped
> * legacy (compressed)
> * legacy (uncompressed)

I have updated to yield the 5 wallet types, improved test coverage and added labels for Legacy types 

Screenshot_1778822242

Comment (Overtorment) [reply tests/integration/import.test.ts:744]: noice!

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @marcosrdz] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/marcosrdz

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @GladosBlueWallet] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/GladosBlueWallet

**Comment by Overtorment** [comment]
@GladosBlueWallet review plz

File: class/wallet-import.ts [hunk 7]
```diff
@@ -382,6 +383,89 @@ const startImport = (
       yield { wallet: legacyWallet };
     }
 
+    yield { progress: 'Private key in hex/base64' };
+
+    // check if text is in hex or base64 format
+    const isHexKey = /^[0-9a-fA-F]{64}$/.test(text);
+    const isBase64Key = /^[A-Za-z0-9+/=]{43,44}$/.test(text);
```
Comment (GladosBlueWallet) [inline class/wallet-import.ts:390]: This regex accepts `=` anywhere, so malformed 43/44-character strings get promoted into the private-key path. Use strict base64 validation here, or random token soup will keep looking like a wallet.

**Review by GladosBlueWallet — COMMENTED** [review]
This is not validation; it's a sieve. Plenty of 43/44-character debris will satisfy this and get treated as a private key candidate before the real parsers even wake up. Tighten the base64 check or the importer will start hallucinating wallets.
